File: finanze/infrastructure/repository/entity/external_entity_repository.py
import json
from datetime import datetime
from typing import Optional
from uuid import UUID

from application.ports.external_entity_port import (
    ExternalEntityPort,
)
from domain.external_entity import (
    ExternalEntity,
    ExternalEntityStatus,
)
from domain.external_integration import ExternalIntegrationId
from infrastructure.repository.db.client import DBClient
from infrastructure.repository.entity.queries import ExternalEntityQueries


def _map_row(row) -> ExternalEntity:
    return ExternalEntity(
        id=UUID(row["id"]),
        entity_id=UUID(row["entity_id"]),
        status=ExternalEntityStatus(row["status"]),
        provider=ExternalIntegrationId(row["provider"]),
        date=datetime.fromisoformat(row["date"]),
        provider_instance_id=row["provider_instance_id"],
        payload=json.loads(row["payload"]) if row["payload"] else None,
    )


class ExternalEntityRepository(ExternalEntityPort):
    def __init__(self, client: DBClient):
        self._db_client = client

    async def upsert(self, ee: ExternalEntity):
        async with self._db_client.tx() as cursor:
            await cursor.execute(
                ExternalEntityQueries.UPSERT,
                (
                    str(ee.id),
                    str(ee.entity_id),
                    ee.status.value,
                    ee.provider.value,
                    ee.date.isoformat(),
                    ee.provider_instance_id,
                    json.dumps(ee.payload) if ee.payload is not None else None,
                ),
            )

    async def update_status(self, ee_id: UUID, status: ExternalEntityStatus):
        async with self._db_client.tx() as cursor:
            await cursor.execute(
                ExternalEntityQueries.UPDATE_STATUS,
                (status.value, str(ee_id)),
            )

    async def get_by_id(self, ee_id: UUID) -> Optional[ExternalEntity]:
        async with self._db_client.read() as cursor:
            await cursor.execute(
                ExternalEntityQueries.GET_BY_ID,
                (str(ee_id),),
            )
            row = await cursor.fetchone()
            return _map_row(row) if row else None

    async def get_by_entity_id(self, entity_id: UUID) -> Optional[ExternalEntity]:
        async with self._db_client.read() as cursor:
            await cursor.execute(
                ExternalEntityQueries.GET_BY_ENTITY_ID,
                (str(entity_id),),
            )
            row = await cursor.fetchone()
            return _map_row(row) if row else None

    async def delete_by_id(self, ee_id: UUID):
        async with self._db_client.tx() as cursor:
            await cursor.execute(
                ExternalEntityQueries.DELETE_BY_ID,
                (str(ee_id),),
            )

    async def get_all(self) -> list[ExternalEntity]:
        async with self._db_client.read() as cursor:
            await cursor.execute(ExternalEntityQueries.GET_ALL)
            rows = await cursor.fetchall()
            return [_map_row(row) for row in rows]

File: mobile/src/services/database/repositories/externalEntityRepository.ts
import { DataManager } from "../dataManager"
import { ExternalEntity, ExternalEntityStatus } from "@/domain"
import { ExternalEntityPort } from "@/application/ports"
import { ExternalEntityQueries } from "./queries"

export class ExternalEntityRepository implements ExternalEntityPort {
  constructor(private client: DataManager) {}

  async upsert(ee: ExternalEntity): Promise<void> {
    await this.client.execute(ExternalEntityQueries.UPSERT, [
      ee.id,
      ee.entityId,
      ee.status,
      ee.provider,
      ee.date ?? null,
      ee.providerInstanceId ?? null,
      ee.payload != null ? JSON.stringify(ee.payload) : null,
    ])
  }

  async updateStatus(
    eeId: string,
    status: ExternalEntityStatus,
  ): Promise<void> {
    await this.client.execute(ExternalEntityQueries.UPDATE_STATUS, [
      status,
      eeId,
    ])
  }

  async getById(eeId: string): Promise<ExternalEntity | null> {
    try {
      const result = await this.client.query<any>(
        ExternalEntityQueries.GET_BY_ID,
        [eeId],
      )
      if (result.rows.length === 0) return null
      return this.mapRow(result.rows[0])
    } catch {
      return null
    }
  }

  async getByEntityId(entityId: string): Promise<ExternalEntity | null> {
    try {
      const result = await this.client.query<any>(
        ExternalEntityQueries.GET_BY_ENTITY_ID,
        [entityId],
      )

      if (result.rows.length === 0) return null

      return this.mapRow(result.rows[0])
    } catch {
      return null
    }
  }

  async deleteById(eeId: string): Promise<void> {
    await this.client.execute(ExternalEntityQueries.DELETE_BY_ID, [eeId])
  }

  async getAll(): Promise<ExternalEntity[]> {
    try {
      const result = await this.client.query<any>(ExternalEntityQueries.GET_ALL)
      return result.rows.map((r: any) => this.mapRow(r))
    } catch {
      return []
    }
  }

  private mapRow(row: any): ExternalEntity {
    let payload: Record<string, unknown> | null = null
    try {
      payload = row.payload ? JSON.parse(row.payload) : null
    } catch {
      payload = null
    }

    return {
      id: row.id,
      entityId: row.entity_id,
      status: row.status as ExternalEntityStatus,
      provider: row.provider,
      date: row.date ?? null,
      providerInstanceId: row.provider_instance_id ?? null,
      payload,
    }
  }
}
